### ISSUE TITLE
JSRuntime.Invoke APIs aren't correctly annotated for trimming

### ISSUE THREAD

**Issue by pranavkm** [dotnet/aspnetcore#39839]
Follow up to https://github.com/dotnet/aspnetcore/pull/39838. JSRuntime's API (excluding the unmarshalled ones) use JSON serialization to serialize the args array and as such are subject to being trimmed away. S.T.Js API for its serialization are annotated with RequiresUnreferencedCode with a recommendation to use the source-generator based overload to avoid this.

Blazor's APIs suppress STJ's warnings (since it happens deep in its bowels) and don't have a trimmer safe alternative. Perhaps one option is to annotate all of the JSRuntime APIs with `RequiresUnferencedCode` and add an JSRuntime.InvokeAsync overload that accepts exactly one argument a `JsonTypeInfo` to go with it. Something like so:

```C#
TValue IJSRuntime.InvokeAsync<TArg, TValue>(string method, TArg arg, JsonTypeInfo<TArg> t1,  JsonTypeInfo<TValue> t2, CancellationToken cancellationToken)
```

At the very least, we could require that all code in the framework uses this API for it's interop.

**Comment by pranavkm** [comment]
FYI @TanayParikh

**Comment by pavelsavara** [comment]
> @pavelsavara would this still be applicable once we update to using the new JS interop with [JSImportAttribute] to make it CSP friendly?
> 
> [#37787 (comment)](https://github.com/dotnet/aspnetcore/issues/37787#issuecomment-1238183412)

Yes, those are separate problems. Solving internal client side interop will fix CSP of Blazor as a framework.
This is problem with `IJSRuntime` which will not go away by solving CSP.

If users adopted new runtime interop with `JSImportAttribute` instead of `IJSRuntime` it would help. 
But `JSImportAttribute` is not available on server side (hybrid Blazor), so it's not complete solution.

Adding new trimming friendly signatures to the `IJSRuntime` is way to go. And maybe marking the old signatures obsolete ?

I think that implementing Roslyn code analyzer to hint or force users to pass `JsonTypeInfo` or `JsonSerializerContext` would be good.

**Comment by yugabe** [comment]
Also wanted to mention there are some incorrect suppressions too, which can and do break functionality. I found one here, but I slightly remember more void-returning ones having this mistake:
https://github.com/dotnet/aspnetcore/blob/1c385f463e488dc9a280fc24352740e919c6bc3f/src/JSInterop/Microsoft.JSInterop/src/JSInProcessObjectReferenceExtensions.cs#L20

The justification states "The method returns void, so nothing is deserialized.". Although the method indeed returns void, but the `params object?[]? args` arguments are passed to the method unchecked, which will be serialized dynamically. This can break if any of the arguments are trimmed away. It's also quite important, as other libraries do in fact depend on these APIs, like Msal, which has had its fair share of problems regarding trimming too.

A generic typed alternative would go a long way.

@pavelsavara I started to migrate to the new `JSImport` API, but it's not on feature parity with the current Blazor one which makes it very hard to migrate to. Disregarding JSON serialization (I understand it can be easily misused and can be a bottleneck), there is the in-box `ElementReference` and `IJSObjectReference` scenarios, the lack of the latter in the new API especially makes it hard to model JavaScript object graphs correctly, as there is no way to invoke functions on `JSObject` references (at least I haven't found any), and `ElementReference` can't be efficiently migrated to `JSImport` APIs either. It's a shame, as I saw some 30-50 times increase in some benchmarks. I think it would also be worth considering opening up some internal APIs, like `System.Runtime.InteropServices.JavaScript.JSFunctionBinding` to let end users call methods dynamically (instead of making custom workarounds which take "methodName" and other parameters).

> Adding new trimming friendly signatures to the IJSRuntime is way to go. And maybe marking the old signatures obsolete ?

👍👍👍

**Comment by pavelsavara** [comment]
> there is the in-box `ElementReference` and `IJSObjectReference` scenarios, the lack of the latter in the new API especially makes it hard to model JavaScript object graphs correctly, as there is no way to invoke functions on `JSObject` references

That's there on purpose, making OOP calls over interop boundary is usually bad design because those tend to be chatty interactions. Especially talking to DOM. Instead you are better off if you aggregate the logic into single JS method and call that. 

Also for first version of the interop we limitted the scope. We started with low level and fast scenarios. We will collect feedback about use cases and see if they could be solved by code generators developed by the community or if there need to be runtime improvements.

That said, what is the specific use case in which you thing it's good idea to call methods on JS instances ?

**Comment by yugabe** [comment]
> That said, what is the specific use case in which you think it's good idea to call methods on JS instances ?

There isn't really a definite "gotcha", but obviously sometimes you need the context of the object to execute a function... That's how object-oriented programming was born. Yes, I can "simulate" object-oriented design by making global/static methods in JavaScript and passing the context object as its first parameter. But that defeats the whole purpose of object-oriented features in both JavaScript and C#. I'd rather not bind to JavaScript's `this` either, there's been quite a few years since that went out of fashion.

I'm not saying it's effectively a "good idea" to call methods on instances, but for example I create typed proxy instances for JavaScript managed instances for third party libraries. I store the Bootstrap plugin "ScrollSpy" instance instantiated along with my Blazor component, so their lifecycles are synchronized, and event handling can be delegated this way from the component through to the third party library. I could mess with lookup dictionaries on both sides and generated IDs, but that's just me writing the framework for something that could be in it anyway.

> That's there on purpose, making OOP calls over interop boundary is usually bad design because those tend to be chatty interactions. Especially talking to DOM.

This is exactly right. But I will call the interaction anyways (because I need to), whether the framework provides the API for it or not. It's just harder to do and more inconvenient if there is no way to call methods directly on objects. If it's an artificial limitation, then it's just making harder for users to generalize code, or write dependent libraries.

If the user calls DOM queries with each method call in a loop, that's on them, their mistake. But preventing calling methods in object instances won't prevent this in any way, in my opinion. `System.Runtime.InteropServices.JavaScript.JSFunctionBinding` being internal won't stop me from forking it either, it's just a lot more troublesome and it would be nice if I didn't have to.

**Comment by nw5428** [comment]
How about just allow passing a JsonSerializerContext to make it easy?

**Comment by pavelsavara** [comment]
> How about just allow passing a JsonSerializerContext to make it easy?

Thank you, could you please elaborate ? 

What would be the shape of the API ?
Would it allow us to eventually disable dynamic JSON for applications which adopt the new API ? (trim reflection codebase away)
Would it be possible to create Roslyn generator/analyzer that would do the right thing on call sites ?
What would be the download size impact for WASM ?
Demo/prototype would be awesome!

**Comment by nw5428** [comment]
await JS.InvokeVoidAsync("doSomething", myObject, FooBar.Default.Foo);  where FooBar.Default.Foo is the source generated strongly typed delegate reference.  You can of course serialize manually and pass it in as a string, but then you also have to deserialize manually on the javascript side.  

InvokeVoidAsync does currently do params object?[]? args, which makes it more complicated but you could do something like 
InvokeVoidAsync <TValue>(IJSObjectReference jsObjectReference,  TValue value, JsonTypeInfo<TValue> jsonTypeInfo, CancellationToken cancellationToken = default)

So you lose the ability to pass multiple parameters but I think this covers a lot of use cases being able to send a single object.

**Comment by ivanjx** [comment]
why not do something like `TypeInfoResolverChain` from web api.

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-10.0#minimal-apis-and-json-payloads
